packages/core-ethereum/crates/core-ethereum-actions/src/transaction_queue.rs

```diff
@@ -0,0 +1,398 @@
+use async_lock::RwLock;
+use async_std::channel::{bounded, Receiver, Sender};
+use async_trait::async_trait;
+use core_crypto::types::Hash;
+use core_ethereum_db::traits::HoprCoreEthereumDbActions;
+use core_types::acknowledgement::AcknowledgedTicket;
+use core_types::acknowledgement::AcknowledgedTicketStatus::BeingRedeemed;
+use core_types::channels::ChannelStatus::{Closed, Open, PendingToClose};
+use core_types::channels::{ChannelEntry, ChannelStatus};
+use std::fmt::{Display, Formatter};
+use std::sync::Arc;
+use utils_log::{debug, error, info, warn};
+use utils_types::primitives::{Address, Balance};
+
+use crate::errors::CoreEthereumActionsError::TransactionSubmissionFailed;
+use crate::errors::Result;
+use crate::transaction_queue::TransactionResult::{Failure, RedeemTicket};
+
+#[cfg(any(not(feature = "wasm"), test))]
+use async_std::task::spawn_local;
+
+#[cfg(all(feature = "wasm", not(test)))]
+use wasm_bindgen_futures::spawn_local;
+
+/// Enumerates all possible outgoing transactions
+#[derive(Clone, PartialEq, Debug)]
+pub enum Transaction {
+    /// Redeem the given acknowledged ticket
+    RedeemTicket(AcknowledgedTicket),
+
+    /// Open channel to the given destination with the given stake
+    OpenChannel(Address, Balance),
+
+    /// Fund channel with the given ID and amount
+    FundChannel(ChannelEntry, Balance),
+
+    /// Close channel with the given source and destination
+    CloseChannel(ChannelEntry),
+
+    /// Withdraw given balance to the given address
+    Withdraw(Address, Balance),
+}
+
+impl Display for Transaction {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Transaction::RedeemTicket(ack) => write!(f, "redeem tx of {ack}"),
+            Transaction::OpenChannel(dst, amount) => write!(f, "open channel tx to {dst} with {amount}"),
+            Transaction::FundChannel(channel, amount) => write!(
+                f,
+                "fund channel tx for channel from {} to {} with {amount}",
+                channel.source, channel.destination
+            ),
+            Transaction::CloseChannel(channel) => write!(
+                f,
+                "closure tx of channel from {} to {}",
+                channel.source, channel.destination
+            ),
+            Transaction::Withdraw(dst, amount) => write!(f, "withdraw tx of {amount} to {dst}"),
+        }
+    }
+}
+
+/// Implements execution of each `Transaction` and also **awaits** its confirmation.
+/// Each operation must return the corresponding `TransactionResult` variant or `Failure`.
+#[cfg_attr(test, mockall::automock)]
+#[async_trait(? Send)]
```

Comment:
DB was complex, these objects should likely be Send.

Reply:
I had it `Send` before, but unfortunately they cannot be, as long as we're calling them from wasm-bound async functions.